Rover/JetsonGStreamerServer/gserver.cpp

```diff
@@ -0,0 +1,227 @@
+// udpsink host=192.168.0.5 port=5555
+GstElement * start_device(GstDevice *dev, const char *client) 
+{
+  GstStructure *str;
+  gchar *name, *cl;
+  GstElement *pipeline, *source, *depay, *encoder, *parse, *sink;
+  GstElement *s_cap, *e_cap, *scale, *scale_cap;
+  GstCaps *s_cap_unapplied, *e_cap_unapplied, *scale_cap_unapplied;
+
+  name = gst_device_get_display_name(dev);
+  cl = gst_device_get_device_class(dev);
+  str = gst_device_get_properties(dev);
+  printf("device name: %s\n", name);
+  printf("device class: %s\n", cl);
+  g_print("device path: %s\n", gst_structure_get_string(str, "device.path")); 
+  
+  // setting elements of pipeline
+  pipeline = gst_pipeline_new(gst_device_get_display_name(dev));
+  source = gst_element_factory_make("v4l2src", NULL);
+  s_cap = gst_element_factory_make("capsfilter", NULL);
+  e_cap = gst_element_factory_make("capsfilter", NULL);
+  depay = gst_element_factory_make("rtph264pay", NULL);
+  encoder = gst_element_factory_make("omxh264enc", NULL);
+  parse = gst_element_factory_make("h264parse", NULL);
+  sink = gst_element_factory_make("udpsink", NULL); 
+  scale = gst_element_factory_make("videoscale", NULL);
+  scale_cap = gst_element_factory_make("capsfilter", NULL);
+
+  // options for the source
+  g_object_set(G_OBJECT(source), "device", gst_structure_get_string(str, "device.path")
+                                                                                , NULL);
+  std::unordered_map<std::string, std::tuple<gint, gint, gint, gint, gint> > table;
+  table["RICOH THETA S"] = std::make_tuple(1280, 720, 5555, 640, 360);
+  table["ZED"] = std::make_tuple(3840, 1080, 5556, 960, 270);
+  table["USB 2.0 Camera"] = std::make_tuple(1280, 720, 5557, 640, 360);
+ 
+  
+  s_cap_unapplied = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING, "I420",
+                              "width", G_TYPE_INT, std::get<0>(table[name]), "height", 
+                               G_TYPE_INT, std::get<1>(table[name]), NULL);
+  g_object_set(G_OBJECT(s_cap), "caps", s_cap_unapplied, NULL);
+
+  // options for the encoder
+  e_cap_unapplied = gst_caps_new_simple("video/x-h264", "stream-format", 
+                                     G_TYPE_STRING, "byte-stream", NULL);
+  g_object_set(G_OBJECT(e_cap), "caps", e_cap_unapplied, NULL);
+
+  // options for the videoscalar
+  scale_cap_unapplied = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING,"I420", 
+                                    "width", G_TYPE_INT, std::get<3>(table[name]), 
+                                    "height", G_TYPE_INT, std::get<4>(table[name]), NULL);
+ 
+  g_object_set(G_OBJECT(scale_cap), "caps", scale_cap_unapplied, NULL);
+
+  // options for the sink
+  g_object_set(G_OBJECT(sink), "host", client, "port", std::get<2>(table[name]), NULL);
+  if (!pipeline || !source || !s_cap || !e_cap || !depay || !encoder || !parse || !sink || !scale || !scale_cap) 
+  {
+    g_printerr ("One element could not be created. SAD! \n");
```

Comment:
This error message is not very helpful, but it's probably fine if it doesn't fail often and you happen to like using gdb